File: neo/src/dal/models/SongWithMeta.ts
import * as t from 'io-ts';

export const codec = t.type({
  id: t.number,
  artists: t.array(t.string),
  title: t.string,
  albumArt: t.union([t.string, t.null]),
  spotifyId: t.union([t.string, t.null]),
  appleMusicId: t.union([t.string, t.null]),
  appleMusicUrl: t.union([t.string, t.null]),
  isrcId: t.union([t.string, t.null]),
  album: t.union([t.string, t.null]),
  isLiked: t.boolean,
  // XXX: This is a string and not a number because SELECT COUNT(*) in postgres
  // returns a bigint, which node-postgres serializes as a string
  likeCount: t.string,
});

export type Model = t.TypeOf<typeof codec>;

File: src/model/SongWithMeta.kt
package club.jambuds.model

import java.time.Instant

data class SongWithMeta(
    // TODO: Split these fields into "Song" resource for songs w/o meta attached
    val id: Int?,
    val createdAt: Instant,
    val title: String,
    val artists: List<String>,
    val album: String?,
    val albumArt: String?,
    val spotifyId: String?,
    val isrcId: String?,
    val appleMusicId: String?,
    val appleMusicUrl: String?,
    // Meta
    val likeCount: Int,
    val isLiked: Boolean
)
